core/drivers/crypto/hisilicon/hisi_qm.c

```diff
@@ -606,19 +606,25 @@ static enum hisi_drv_status qm_cqc_cfg(struct hisi_qp *qp)
 struct hisi_qp *hisi_qm_create_qp(struct hisi_qm *qm, uint8_t sq_type)
 {
 	struct hisi_qp *qp = NULL;
+	int cur_idx = 0;
+	uint32_t i = 0;
 
 	mutex_lock(&qm->qp_lock);
 	if (qm->qp_in_used == qm->qp_num) {
 		EMSG("All %"PRIu32" queues of QM are busy", qm->qp_num);
 		goto err_proc;
 	}
 
-	if (qm->qp_idx == qm->qp_num - 1)
-		qm->qp_idx = 0;
-	else
-		qm->qp_idx++;
+	for (i = 0; i < qm->qp_num; i++) {
+		cur_idx = (qm->qp_idx + i) % qm->qp_num;
+		if (qm->qp_array[cur_idx].used == 0) {
+			qm->qp_array[cur_idx].used = true;
+			qm->qp_idx = cur_idx + 1;
+			break;
+		}
+	}
```

Comment:
panic if all used? or return NULL? This makes last tested index being used.
unlock/wait/relock, cond_var?

Reply:
![image](https://github.com/OP-TEE/optee_os/assets/118411745/2ab9abc5-a739-4945-a77a-07d107189214)
This avoid all used case.

Reply:
right, sorry.